src/relax/transform/legalize_ops.cc

```diff
@@ -152,6 +164,68 @@ class LegalizeMutator : public ExprMutator {
     return NullOpt;
   }
 
+  Expr UpdateVDeviceOutStructInfo(Expr expr, const Call& visited_call) {
+    static const auto& infer_struct_info_map = Op::GetAttrMap<FInferStructInfo>("FInferStructInfo");
+    static const Op& call_tir_op = Op::Get("relax.call_tir");
+    auto* op_node = visited_call->op.as<OpNode>();
+
+    // Not an OpNode
+    if (op_node == nullptr) {
+      return expr;
+    }
+    auto op = GetRef<Op>(op_node);
+
+    if (!infer_struct_info_map.count(op)) {
+      return expr;
+    }
+
+    if (!expr->IsInstance<CallNode>()) {
+      return expr;
+    }
+
+    auto call = Downcast<Call>(expr);
+    if (call->op != call_tir_op) {
+      return expr;
+    }
+
+    StructInfo out_sinfo = call->sinfo_args[0];
+    StructInfo infered_sinfo = infer_struct_info_map[op](visited_call, builder_);
```

Comment:
seems we are mixing legalization and device propagation. Can we leverage structure info before the the post order visit (aka original Call), assuming the result dont change during legalization? so we don't have to redo inference.

Reply:
@srkreddy1238 seems this one is not yet resolved?

Reply:
Forgot to mark. Its resolved.